File: Hard/C++/Largest_Rectangle_in_Histogram(84).cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        
        
        int res=0;
        heights.push_back(0);// to clear the vector at end
        vector<int> index;// store the indexes to get the width easily
        
        for(int i=0;i<heights.size();i++)
        {
            
            while(index.size()>0 and heights[i]<heights[index.back()])
            {
                
                int h=heights[index.back()];// start poping the bars coz we can not streach them further
                
                index.pop_back();
                
                int x=index.size() > 0 ? index.back() : -1;// getting the index
                
                
                res=max(res,h*(i-x-1));// i-x-1 is the width of the current rectangle with height h;
    
            }
            index.push_back(i);
        }
        return res;
    }
};

File: Hard/Java/Largest_Rectangle_in_Histogram(84).java
class Solution {
    public int largestRectangleArea(int[] heights) {
        if(heights == null || heights.length == 0) return 0;
        int n=heights.length;
        Stack<Integer>st=new Stack<>();
        int area=0;
        for(int i=0;i<=n;i++)
        {
            int h = ( i==n ? 0:heights[i] );
            if(st.isEmpty() || h>=heights[st.peek()] )
                st.push(i);
            else
            {
                int tmp=st.pop();
                area=Math.max( area , heights[tmp]*(st.isEmpty()?i:i-st.peek()-1) ) ;
                i--;
            }
        }
        return area;
    }
}